tests/unit/s2n_cookie_test.c

```diff
@@ -0,0 +1,368 @@
+     *# extension to the client (this is an exception to the usual rule that
+     *# the only extensions that may be sent are those that appear in the
+     *# ClientHello).  When sending the new ClientHello, the client MUST copy
+     *# the contents of the extension received in the HelloRetryRequest into
+     *# a "cookie" extension in the new ClientHello.
+     */
+    for (size_t i = 0; i < TEST_COOKIE_COUNT; i++) {
+        DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(server_conn);
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+        DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(server_conn);
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+        struct s2n_test_io_pair io_pair = { 0 };
+        EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+        EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+        /* Force the HRR path */
+        client_conn->security_policy_override = &security_policy_test_tls13_retry;
+
+        /* Force the server to send a cookie */
+        EXPECT_SUCCESS(s2n_dup(&test_cookies[i], &server_conn->cookie));
+
+        /* Negotiate handshake */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+
+        /* Verify HRR path */
+        EXPECT_TRUE(IS_HELLO_RETRY_HANDSHAKE(server_conn));
+        EXPECT_TRUE(IS_HELLO_RETRY_HANDSHAKE(client_conn));
+
+        /* Verify cookies */
+        S2N_BLOB_EXPECT_EQUAL(test_cookies[i], client_conn->cookie);
+        S2N_BLOB_EXPECT_EQUAL(test_cookies[i], server_conn->cookie);
+
+        EXPECT_SUCCESS(s2n_shutdown_test_server_and_client(server_conn, client_conn));
+        EXPECT_SUCCESS(s2n_io_pair_close(&io_pair));
+    }
+
+    /* Self-Talk: Full connection lifecycle with cookies
+     * We try the handshake multiple times with different possible call patterns.
+     */
+    {
+        DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(server_conn);
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+        DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(server_conn);
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+        for (size_t i = 0; i < 100; i++) {
+
+            uint64_t r = 0;
+            EXPECT_OK(s2n_public_random(100, &r));
+            bool hrr = (r % 2) == 0;
+            bool cookie = (r % 3) == 0;
+            size_t cookie_i = i % TEST_COOKIE_COUNT;
+            bool free_handshake = (r % 7) == 0;
```

Comment:
the only down side of this is that these test now become non-deterministic. Can you print out the value of `r` so its possible to repro a broken test case.

Reply:
Good point. Rather than printing all the `r`s, I switched to a pseudo random sequence based on a known, fixed seed.